internal/datasets/subjectsetbytype.go

```diff
@@ -19,24 +19,24 @@ func NewSubjectByTypeSet() *SubjectByTypeSet {
 }
 
 // AddSubjectOf adds the subject found in the given relationship, along with its caveat.
-func (s *SubjectByTypeSet) AddSubjectOf(relationship *core.RelationTuple) error {
-	return s.AddSubject(relationship.Subject, relationship.Caveat)
+func (s *SubjectByTypeSet) AddSubjectOf(relationship tuple.Relationship) error {
+	return s.AddSubject(relationship.Subject, relationship.OptionalCaveat)
 }
 
 // AddConcreteSubject adds a non-caveated subject to the set.
-func (s *SubjectByTypeSet) AddConcreteSubject(subject *core.ObjectAndRelation) error {
+func (s *SubjectByTypeSet) AddConcreteSubject(subject tuple.ObjectAndRelation) error {
 	return s.AddSubject(subject, nil)
 }
 
 // AddSubject adds the specified subject to the set.
-func (s *SubjectByTypeSet) AddSubject(subject *core.ObjectAndRelation, caveat *core.ContextualizedCaveat) error {
-	key := tuple.JoinRelRef(subject.Namespace, subject.Relation)
+func (s *SubjectByTypeSet) AddSubject(subject tuple.ObjectAndRelation, caveat *core.ContextualizedCaveat) error {
```

Comment:
A few reasons:

1. They aren't used nearly as often, so the space savings are minimal
2. Context can be quite large, so copying it around on the stack seems unwise
3. Context is often `nil`, so there won't be any overhead in the common case

Reply:
Thanks for the context. It just feels _odd_ because it won't be clear to the reader why the codebase is left in this odd mixture of structs and protos, and what pattern should be followed as the codebase evolves. From the arguments laid, context being _quite large_ seems like something that would be equally problematic for the heap.

Reply:
Context being large is better to be on the heap: it ensures we're not copying it around; granted, context is not usually super large, but either way, its not used as much